### DISCUSSION TITLE
Static interfaces

### DISCUSSION THREAD

**Discussion by LWChris** [dotnet/csharplang#6048] [Language Ideas]
An interface is a contract on what members to expect. Usually, that contract is applied to an instance and instance methods. But why shouldn't I be able to declare that any instance or even a static class must have a certain set of static members as well?

Definition:
```
public static interface ILogger
{
  static void WriteLine(string message);
}
```

Implementation:
```
public static class ConsoleLogger : ILogger // note: only static interfaces are accepted here
{
  // Implicit implementation
  public static void WriteLine(string message) => System.Console.WriteLine(message);
}

public static class DebugLogger : ILogger
{
  // Explicit implementation
  static void ILogger.WriteLine(string message) => System.Diagnostics.Debug.WriteLine(message);
}
```

Usage:
```
public class Worker<TLogger> where TLogger : ILogger
{
  public async Task DoWork() {
    TLogger.WriteLine("Started"); // works since it is known whichever class goes into the generic will have a matching static method
    await Task.Delay(1000);
    TLogger.WriteLine("Finished");
  }
}

public class LogMessageGenerator
{
  public async Task Produce()
  {
#if DEBUG
    var worker = new Worker<DebugLogger>();
#else
    var worker = new Worker<ConsoleLogger>();
#endif
    await worker.DoWork();
  }
}
```

Note:
I am aware that for 1 method a simple delegate would do fine. But this is a small example to illustrate the use.
Static interfaces could be useful for other scenarios like Create patterns:

```
public interface IProducer
{
  int Result { get; }
}

public static interface ICreateable<out T, TArgs>
{
  static T Create(TArgs arguments);
}

public class Four : IProducer
{
  public int Result => 4;
}

public class MyInt : ICreatable<IProducer, int>, IProducer
{  
  public MyInt(int i) => Result = i;
  public int Result { get; }
  static MyInt ICreatable<MyInt, int>.Create(int i) => new(i);
}

public class Summarizer
{
  private void AddProducer(Func<IProducer> factory);
  public void Add<T>() where T : new(), IProducer => AddProducer(() => new T());
  public void Add<T, TArgs>(TArgs args) where T : ICreatable<T, TArgs>, IProducer => AddProducer(() => T.Create(args));
}
```

**Comment by PathogenDavid** [comment]
See https://github.com/dotnet/csharplang/issues/4436

[This feature is already available in C# previews.](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/static-abstract-interface-methods)

See also [the blog post on generic math](https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/).

**Reply by LWChris** [reply]
I saw/read this right before posting, but I'm not sure it's _quite_ the same.

Declaring static members in normal interfaces allows me to declare that a certain generic type has these static members.

The point in marking the _interface_ with the `static` keyword however is to say that _all_ members of that interface are static, and therefore this particular interface may be safely applied to a `static class`.

Since in #4436 the interface is still a "normal" interface, I think static classes still aren't able to be interfaced, are they?

**Reply by PathogenDavid** [reply]
Oh, sorry. I overlooked that part of your post.

That would be covered by https://github.com/dotnet/csharplang/issues/5783

It was [discussed in LDM back in February](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-02-16.md#static-abstract-interfaces-and-static-classes) and is under consideration, but I don't think it's going to happen for C# 11.

**Reply by LWChris** [reply]
Aha, so my proposal is a hybrid of #4436 and #5783.

Taking the "apply to static classes" part from #5783, and getting rid of a lot of headache they had over what happens with non-static members of interfaces that are applied to static classes, which they also discussed but not settled on yet?

And showing how this already allows for some use cases of the much more complex #4436 proposal. (TBH it's still over my head why we need the `abstract` keyword all over the place, because interface methods have always been conceptually like abstract methods, or later on virtual when they could have default implementations. This has always been the case, without the necessity for keywords).

**Reply by HaloFour** [reply]
@LWChris 

> TBH it's still over my head why we need the abstract keyword all over the place, because interface methods have always been conceptually like abstract methods

The problem was that C# started to allow `static` members on interfaces in C# 8.0, and they were just normal static members which are not virtual.  They needed to distinguish these new virtual static members from existing non-virtual static members.

**Reply by CyrusNajmabadi** [reply]
> or later on virtual when they could have default implementations. This has always been the case, without the necessity for keywords).

The problem is that interface can already have static *non-virtual* methods *today*.  So we needed a way to distinguish between those.

**Reply by CyrusNajmabadi** [reply]
Oh.  @HaloFour answered this.  Sorry, my github hadn't shown that :)

**Comment by HaloFour** [comment]
To be consistent with `static class` I believe all of the members of the interface would also have to be `static`:

```cs
public static interface ILogger
{
  static abstract void WriteLine(string message);
}
```

**Reply by LWChris** [reply]
My notation was coming the "everything before `interface` is implied for all members" route.

Like if you don't have to write

```
**public** interface ILogger
{
  **public** void WriteLine(string message);
}
```

you shouldn't have to write

```
**public static** interface IStaticLogger
{
  **public static** void WriteLine(string message);
}
```

But I do agree though that it is more consistent for implicit vs explicit implementation then (unless explicit implementation also omits the `static` keyword, of which I am 100% not a fan).

```
public class Logger : ILogger
{
  // Implicit: **public** needed
  public void WriteLine...
  
  // Explicit: **public** omitted, looks like in the interface
  void ILogger.WriteLine...
}

public static class static Logger : IStaticLogger
{
  // Implicit: **public static** needed
  public static void WriteLine...
  
  // Explicit: **public static** omitted so it looks like in the interface? Ew!
  void IStaticLogger.WriteLine...
  
  // Explicit: Just **public** omitted. Better.
  // But if this is what the interface should look like, then members in interface should have "static"
  static void IStaticLogger.WriteLine...
}
```

**Reply by HaloFour** [reply]
@LWChris 

That's already not the case:

```cs
interface IFoo {
    void Bar();
}
```

In this situation `IFoo` is `internal` and `Bar` is `public`.  The default accessibility is not transitive, and neither are any other modifiers like `static`.  Otherwise, that would already be the case with `static` classes, and it is not.

**Reply by LWChris** [reply]
I agree. So what's the proper way to go from here (new to discussions)? Should I just "edit" the original post to include the keywords, so it reflects the "best" state? Or is it frowned upon to change a proposal?

**Reply by HaloFour** [reply]
That's up to you.  😄 

I think the exact syntax isn't as important as whether there is value in having a way to apply interfaces to `static` classes, whether that be a new form of interface or not.

**Reply by LWChris** [reply]
> Otherwise, that would already be the case with `static` classes, and it is not.

To be honest, I find it annoying that the static keyword is not transitive in static classes, because I think there's no point in writing "static" before every member if that's literally your only choice anyway.

It makes sense **not** to require the ctor name of generic classes to contain the generics - because of course it's a ctor for this generic class, because that's litereally the only thing your ctor can be for.
Same logic should apply to `static` within `static class` - of course it's a static member, because that's literally the only thing a `static class` can contain.

But I digress... 😅

**Reply by jnm2** [reply]
> To be honest, I find it annoying that the static keyword is not transitive in static classes, because I think there's no point in writing "static" before every member if that's literally your only choice anyway.

If the class becomes non-static, this prevents mistakes, prevents instant loads of compile errors for all usages of the methods, and keeps source control line history cleaner.

**Reply by LWChris** [reply]
Okay, I can see some value/benefit in this. It is certainly less often a use case for me to un-`static` a class than forgetting to write `static`. Source control-wise it would also make sense to say you wanted to make the class non-static, now you go through the methods and verify that those should be still static. But I can accept that the other way around makes sense, too. 🙂
